### PR TITLE
Hvg 893 embark tracking

### PR DESCRIPTION
**Pull request by tarikstafford**

### CODE REVIEW

**Review by sampettersson — DISMISSED** [review]
Neat! Code looks good, maybe we should write a test for this logic though?

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+public extension EmbarkTrackingEvent {
+    static var trackingHandler: (_ Event: EmbarkTrackingEvent) -> Void = { _ in }
+}
+
+internal extension GraphQL.EmbarkExternalRedirectLocation {
+    func trackingEvent(store: EmbarkStore) -> EmbarkTrackingEvent {
+        var properties = [String:String]()
+        if let store = store.revisions.last {
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:21]: maybe we should have a `func getAllValues() -> [String: String]` and that one just goes into revisions and picks the last one. what you think?

Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:21]: Use `getAllValues()` here too?

File: Projects/Embark/Sources/EmbarkStore.swift [hunk 2]
```diff
@@ -36,6 +36,10 @@ class EmbarkStore {
             }
         }
     }
+    
+    func getAllValues() -> [String:String] {
+        return revisions.last ?? [:]
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkStore.swift:41]: We would need to evaluate all computed values here before returning right? Also will we get any performance implications by that 🤔

**Review by sampettersson — CHANGES_REQUESTED** [review]

File: Projects/Embark/Sources/EmbarkState.swift [hunk 3]
```diff
@@ -50,8 +54,18 @@ public struct EmbarkState {
             return computedValues
         } ?? [:]
     }
+    
+    func startTracking() {
+        bag += currentPassageSignal
+                .readOnly()
+                .compactMap { $0?.tracks }
+                .onValue({ (tracks) in
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState.swift:62]: Use the background scheduler here: `on: .background`

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+internal extension EmbarkPassage.Track {
+    func trackingEvent(store: EmbarkStore) -> EmbarkTrackingEvent {
+        return .init(title: self.eventName, properties: properties(store: store))
+    }
+    
+    private func properties(store: EmbarkStore) -> [String:Any] {
+        var properties = Dictionary<String,Any>()
+        if includeAllKeys {
+            properties = properties.merging((store.getAllValues())) { return $1 }
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:34]: You can use `takeRight` here.

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        } else {
+            let storeProperties = store.getAllValues().filter { key, value in
+                return eventKeys.contains(key)
+            }
+            
+            properties = properties.merging(storeProperties, uniquingKeysWith: { return $1 })
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:40]: You can use `takeRight` here.

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        
+        if let customData = customData {
+            properties = properties.merging((customData.toJSONDictionary() ?? [:]), uniquingKeysWith: { return $1 })
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:44]: You can use `takeRight` here.

**Review by sampettersson — CHANGES_REQUESTED** [review]

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+internal extension EmbarkPassage.Track {
+    func trackingEvent(store: EmbarkStore) -> EmbarkTrackingEvent {
+        return .init(title: self.eventName, properties: properties(store: store))
+    }
+    
+    private func properties(store: EmbarkStore) -> [String:Any] {
+        var properties = Dictionary<String,Any>()
+        if includeAllKeys {
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:33]: it's been decided that this is not something we wanna support going forward, due to there being unwanted data sent to Mixpanel, so we can remove this for now.

File: Projects/Embark/Sources/EmbarkState+Tracking.swift [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    }
+}
+
+public extension EmbarkTrackingEvent {
+    static var trackingHandler: (_ Event: EmbarkTrackingEvent) -> Void = { _ in }
+}
+
+internal extension GraphQL.EmbarkExternalRedirectLocation {
+    func trackingEvent(store: EmbarkStore) -> EmbarkTrackingEvent {
+        var properties = store.getAllValues()
```
Comment (sampettersson) [inline Projects/Embark/Sources/EmbarkState+Tracking.swift:20]: Don't send the whole store here.

File: Projects/Embark/Sources/EmbarkState+Tracking.swift (hunk 7 above)
Comment (tarikstafford) [reply Projects/Embark/Sources/EmbarkState+Tracking.swift:33]: But we should still send the entire dictionary on external redirect ?

**Review by sampettersson — APPROVED** [review]
You probably need to merge this against `embark-feature` for the build to run.

**Pull request merged**